src/cilint.ml

```diff
@@ -37,232 +29,126 @@ let m1 = minus_big_int unit_big_int
-  | Small i when n < Nativeint.size - 2 ->
-      let max = 1 lsl (n - 1) in
-      let truncmax = 1 lsl n in
-     let bits = i land (truncmax - 1) in
-      let tval =
-    (* check if the n-th bit is 1... *)
-	if bits < max then
-	  bits
-	else
-    (* and fill with 1 bits on the left if it is *)
-	  bits - truncmax
-      in
-      let trunc =
-	if i >= max || i < -max then
-	  if i >= truncmax then
-	    BitTruncation
+  let max = power_int_positive_int 2 (n - 1) in
+  let truncmax = power_int_positive_int 2 n in
+  let bits = mod_big_int c truncmax in
+  let tval = if lt_big_int bits max then
+	    bits
 	  else
-	    ValueTruncation
-	else
-	  NoTruncation
-      in Small tval, trunc
-  | _ ->
-      let b = big_int_of_cilint c in
-      let max = power_int_positive_int 2 (n - 1) in
-      let truncmax = power_int_positive_int 2 n in
-      let bits = mod_big_int b truncmax in
-      let tval =
-	if lt_big_int bits max then
-	  bits
-	else
-	  sub_big_int bits truncmax
-      in
-      let trunc =
-	if ge_big_int b max || lt_big_int b (minus_big_int max) then
-	  if ge_big_int b truncmax then
-	    BitTruncation
-	  else
-	    ValueTruncation
-	else
-	  NoTruncation
-      in cilint_of_big_int tval, trunc
+	    sub_big_int bits truncmax
+  in
+  let trunc =
+    if ge_big_int c max || lt_big_int c (minus_big_int max) then
+      if ge_big_int c truncmax then
+        BitTruncation
+      else
+        ValueTruncation
+    else
+      NoTruncation
+  in
+    tval, trunc
 
 let truncate_unsigned_cilint (c:cilint) (n:int) : cilint * truncation =
-  match c with
-  | Small i when i > 0 && n >= Nativeint.size - 2 -> Small i, NoTruncation
-  | Small i when n < Nativeint.size - 2 ->
-      let max = 1 lsl (n - 1) in
-      let truncmax = 1 lsl n in
-      let bits = i land (truncmax - 1) in
-      let trunc =
-	if i >= truncmax || i < 0 then
-	  if i < -max then
-	    BitTruncation
-	  else
-	    ValueTruncation
-	else
-	  NoTruncation
-      in Small bits, trunc
-  | _ ->
-      let b = big_int_of_cilint c in
-      let max = power_int_positive_int 2 (n - 1) in
-      let truncmax = power_int_positive_int 2 n in
-      let bits = mod_big_int b truncmax in
-      let trunc =
-	if ge_big_int b truncmax || lt_big_int b zero_big_int then
-	  if lt_big_int b (minus_big_int max) then
-	    BitTruncation
+  let max = power_int_positive_int 2 (n - 1) in
+  let truncmax = power_int_positive_int 2 n in
+  let bits = mod_big_int c truncmax in
+  let trunc =
+    if ge_big_int c truncmax || lt_big_int c zero_big_int then
+      if lt_big_int c (minus_big_int max) then
+        BitTruncation
+      else
+        ValueTruncation
 	  else
-	    ValueTruncation
-	else
-	  NoTruncation
-      in cilint_of_big_int bits, trunc
+	    NoTruncation
+  in
+  bits, trunc
 
-let is_int_cilint (c:cilint) : bool =
-  match c with
-  | Small _ -> true
-  | Big b -> is_int_big_int b
+let is_int_cilint (c:cilint) : bool = is_int_big_int c
```

Comment:
What is this used for? We shouldn't care whether Zarith internally has a small or big thing.

Reply:
This is not about what Zarith has internally but about whether the number could fit into an `int`. There are some functions that check this in `cil.ml` to throw an error otherwise and continue with computation on `int` otherwise.